Readme (#5)
LBHackney-IT/mtfh-reporting-data-listener: README.md

# MTFH Reporting Data Listener

This listener implements an AWS Lambda Function that recieves messages when any updates are made to the TenureInformationApi and sends the updates to a Kafka queue. 

Here is the process on how the data is saved to Kafka:

> 1. When one of the events in from [this switch statement][usecase-factory] is raised then this listener is triggered

Reviewer: ```suggestion
1. When one of the events in [this switch statement][usecase-factory] is raised then this listener is triggered
```

> 2. The listener calls the [TenureInformationApi][tenure-api-github] using a Shared Nuget Package to ensure that the Tenure sent through in the message exists and to get the details of the tenure. [The ReadME][Api-Gateway] for this package explains how the listener calls the TenureInformationApi.

Reviewer 2: Hey Humairaa - as this is the README for the whole listener, can we make the descsription generic and not specific to Tenure information API? We can have a separate section that will show what use cases are currently implemeneted (e.g. Tenure updated)

Author: Sure, I can make that change

     - If the Tenure doesn't exist then the listener throws an Exception 
3. The listener will then get the schema from AWS Glue Schema registry managed in [this repository][ADD LINK HERE]

> 4. Then using this schema, an AVRO generic record is created holding the tenure details retrieved from the tenure API.

Reviewer 2: It might be easier if we add titles to the sections - e.g. this could be 'Converting the data to AVRO format'

> 4. Then using this schema an AVRO generic record is created using holding the tenure details retrieved from the tenure API.

Reviewer: ```suggestion
4. Then using this schema, an AVRO generic record is created holding the tenure details retrieved from the tenure API.
```

>    Kafka only accepts the following data type; `byte[], Bytebuffer, Double, Integer, Long, String`. If you need to send through other data type you will first need to serialize the data into one of these types. Below are some code examples of how we have done this.

Reviewer 3: ```suggestion
   Kafka only accepts the following data types; `byte[], Bytebuffer, Double, Integer, Long, String`. If you need to send through other data types you will first need to serialize the data into one of these types. Below are some code examples of how we have done this.
```

   **Nullable types (Union)**:

   Here we check whether the field value is null, assign null if it is.

>    If not, remove the nullable part from the schema and continue as normal.

Reviewer 2: Can we add a short info about why this is needed?

            jsonSchema.TryGetProperty("items", out var itemsSchemaJson);
            return Schema.Parse(itemsSchemaJson.ToString());
        }
    ``` 

    Once the data types have been converted you can create the Generic Record by adding each value to the record, this can be done through the use of a for loop & reflection:

    ```csharp

        public GenericRecord PopulateFields(object item, Schema schema)
        {
            var record = new GenericRecord((RecordSchema) schema);
            ((RecordSchema) schema).Fields.ForEach(field =>
            {
                PropertyInfo propInfo = item.GetType().GetProperty(field.Name);
                if (propInfo == null)
                {
                    Console.WriteLine($"Field name: {field.Name} not found in {item}");
                    return;
                }

                var fieldValue = propInfo.GetValue(item);
                var fieldSchema = field.Schema;
                var fieldType = field.Schema.Tag;

                //Add any data type conversion required here

                record.Add(field.Name, fieldValue);
            });

            return record;
        }

    ```

> 6. Lastly the record created above is then sent through to Kafka.

Reviewer 2: Maybe separate this into two sections with the following titles: 
Serializing the data

Sending the data to Kafka

>    In order to send the data through to Kafka a SchemaRegistryClient is required. SchemaRegistery are seperate from your Kafka brokers.The Kafka Producers publish the data to Kafka topics and communicates with the Schema Registry to send and recieve schemas that describe the data models for the messages simultaneously. Hence the SchemaRegistry is used to serialize the message and then save the serialize message to Kafka. 

Reviewer 3: ```suggestion
   In order to send the data through to Kafka a SchemaRegistryClient is required. SchemaRegistery are seperate from your Kafka brokers. The Kafka Producers publish the data to Kafka topics and communicates with the Schema Registry to send and receive schemas that describe the data models for the messages simultaneously. Hence the SchemaRegistry is used to serialize the message and then save the serialize message to Kafka. 
```

>    We then use the `Producer.Flush(TimeSpan.FromSeconds(10))` to immediately send through the message to Kafka. There is a 10 seconds maximum timeout, this wouldn't hang the listener for 10 seconds every time.Instead, It will wait until either all the messages have been sent or 10 seconds pass. If we remove the 10 seconds then it would wait forever until all the messages have been sent, which we don't want as if there is an issue, it could just hang until the lambda timeout.

Reviewer 3: ```suggestion
   We then use the `Producer.Flush(TimeSpan.FromSeconds(10))` to immediately send through the message to Kafka. There is a 10 seconds maximum timeout on the listener. This would mean that it will wait until either all the messages have been sent or 10 seconds have passed. If a maximum timeout is not set then it would wait until all the messages have been sent, and in the event there is an issue whilst sending messages, it will likely hang until the lambda timeout which we don't want.
```
